Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/

cdumez opened the pull request:
#### d3c05474b8c2feaf1a3aa51934a70a13769bf622
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/
<a href="https://bugs.webkit.org/show_bug.cgi?id=283201">https://bugs.webkit.org/show_bug.cgi?id=283201</a>

Reviewed by Darin Adler.

* Source/WebCore/loader/FormSubmission.cpp:
(WebCore::appendMailtoPostFormDataToURL):
* Source/WebCore/loader/ResourceCryptographicDigest.cpp:
(WebCore::parseCryptographicDigestImpl):
(WebCore::parseEncodedCryptographicDigestImpl):
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::shouldDeferUpdateImageData const):
* Source/WebCore/page/PerformanceObserver.cpp:
(WebCore::PerformanceObserver::observe):
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::parseDirective):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::parse):
(WebCore::ContentSecurityPolicySourceList::parseSource):
(WebCore::ContentSecurityPolicySourceList::parseScheme):
(WebCore::ContentSecurityPolicySourceList::parseHost):
(WebCore::ContentSecurityPolicySourceList::parsePath):
(WebCore::ContentSecurityPolicySourceList::parsePort):
(WebCore::ContentSecurityPolicySourceList::parseNonceSource):

Canonical link: <a href="https://commits.webkit.org/286696@main">https://commits.webkit.org/286696@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c346e74c94cc080a55720fd6c4aaceb5a4ee543

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76799 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55834 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29705 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81332 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28075 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78916 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64976 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4127 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60180 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18270 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79866 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50129 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65933 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40499 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23427 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26399 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68648 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23757 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82776 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4175 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2767 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68465 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4328 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65905 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67718 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16883 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11694 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9775 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4122 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/6933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4145 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7575 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5903 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/page/PerformanceObserver.cpp:
```diff
@@ -93,11 +91,9 @@ ExceptionOr<void> PerformanceObserver::observe(Init&& init)
-            std::inplace_merge(begin, oldEnd, end, PerformanceEntry::startTimeCompareLessThan);
+            auto entriesToDeliver = m_entriesToDeliver.mutableSpan();
+            std::stable_sort(entriesToDeliver.subspan(oldSize).begin(), entriesToDeliver.end(), PerformanceEntry::startTimeCompareLessThan);
+            std::inplace_merge(entriesToDeliver.begin(), entriesToDeliver.subspan(oldSize).begin(), entriesToDeliver.end(), PerformanceEntry::startTimeCompareLessThan);
```
I think this was easier to read with the 3 local variables; it’s no fun to have the `subspan` expression repeated twice. Would you consider bringing them back?

Is `entriesToDeliver.subspan(oldSize).begin()` safer than `entriesToDeliver.begin() + oldSize`? I would be surprised if span iterator operations were unchecked. Separately, it also seems unfortunate that we can’t make his safer by making `Vector` iterators checked instead of introducing use of a span.

cdumez replied:
> I would be surprised if span iterator operations were unchecked.

Then be surprised :) cc @geoffreygaren.
We may need to create our own WTF::Span so we get checked iterators, libcpp couldn't do it due to ABI compatibility iirc.

cdumez replied:
Vector iterators are currently raw pointers and not safe either. We'll likely want to make them safe as well though, I agree.

darinadler replied:
I don’t want to stand in the way, but I am not clear on the value of the current change. If `Vector` or `span` iterators were checked then this code would be significantly safer. If not, making just the `subspan` call do a range check doesn’t seem all that valuable, and it’s a significant code change. I suppose we need to do something to make the validation checker happy, but it seems a smaller change could possibly accomplish that?

geoffreygaren replied:
Unfortunately, libc++ span iterators are in fact unchecked in user space, and there's no plan to fix that for the time being.

Making Vector iterator checked is an option. (And we should totally do that, or at minimum mark Vector iterator ++ WTF_UNSAFE_BUFFER_USAGE as soon as possible.)

Making a checked span iterator wrapper is also an option. Note that we don't necessarily need to replace all use of span with a new WTF::Span. We can just use a discipline of doing boundsCheckedBegin(span) instead of span.begin(), and fill out more features in BoundsCheckedIterator as we need them. And maybe pick a shorter name while we're at it.

(libc++ is working on adding [[clang::unsafe_buffer_usage]] to span iterators in user space, so we do have a solution to using an unsafe span iterator by accident.)

darinadler replied:
I guess the main thing I am wondering about is how we align the unsafe checks with the strategy. I’m surprised that we can make the checker happy with these changes.

darinadler replied:
If we have a checker that can check when it’s needed, than I think `saferBegin(x)` would be fine. For both `Vector` and `span`.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -518,42 +516,42 @@ template<typename CharacterType> auto ContentSecurityPolicyDirectiveList::parseD
 
     if (!buffer.atEnd()) {
-        m_policy.reportInvalidDirectiveValueCharacter(name, String({ valueBegin, buffer.end() }));
+        m_policy.reportInvalidDirectiveValueCharacter(name, String(valueBegin));
```
I don’t think we need `String()` here; if I am right we should omit it.

darinadler commented on Source/WebCore/loader/ResourceCryptographicDigest.cpp:
```diff
@@ -104,15 +102,15 @@ template<typename CharacterType> static std::optional<EncodedResourceCryptograph
         return std::nullopt;
 
-    return EncodedResourceCryptographicDigest { *algorithm, String({ beginHashValue, buffer.position() }) };
+    return EncodedResourceCryptographicDigest { *algorithm, String(beginHashValue.first(buffer.position() - beginHashValue.data())) };
```
I don’t think we need `String()` here; if I am right we should omit it.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -518,42 +516,42 @@ template<typename CharacterType> auto ContentSecurityPolicyDirectiveList::parseD
+    if (nameBegin.data() == buffer.position()) {
         skipWhile<isNotASCIISpace>(buffer);
-        m_policy.reportUnsupportedDirective(String({ nameBegin, buffer.position() }));
+        m_policy.reportUnsupportedDirective(String(nameBegin.first(buffer.position() - nameBegin.data())));
```
I don’t think we need `String()` here; if I am right we should omit it.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -518,42 +516,42 @@ template<typename CharacterType> auto ContentSecurityPolicyDirectiveList::parseD
     if (!skipExactly<isUnicodeCompatibleASCIIWhitespace>(buffer)) {
         skipWhile<isNotASCIISpace>(buffer);
-        m_policy.reportUnsupportedDirective(String({ nameBegin, buffer.position() }));
+        m_policy.reportUnsupportedDirective(String(nameBegin.first(buffer.position() - nameBegin.data())));
```
I don’t think we need `String()` here; if I am right we should omit it.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
```diff
@@ -270,7 +268,7 @@ template<typename CharacterType> void ContentSecurityPolicySourceList::parse(Str
                 m_list.append(ContentSecurityPolicySource(m_policy, source->scheme.convertToASCIILowercase(), source->host.value.toString(), source->port.value, source->path, source->host.hasWildcard, source->port.hasWildcard, IsSelfSource::No));
         } else
-            m_policy.reportInvalidSourceExpression(m_directiveName, String({ beginSource, buffer.position() }));
+            m_policy.reportInvalidSourceExpression(m_directiveName, String(beginSource.first(buffer.position() - beginSource.data())));
```
I don’t think we need `String()` here; if I am right we should omit it.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
```diff
@@ -515,17 +513,17 @@ template<typename CharacterType> String ContentSecurityPolicySourceList::parsePa
     //                ^                               ^
     if (buffer.hasCharactersRemaining())
-        m_policy.reportInvalidPathCharacter(m_directiveName, String({ begin, buffer.end() }), *buffer);
+        m_policy.reportInvalidPathCharacter(m_directiveName, String(begin), *buffer);
```
I don’t think we need `String()` here; if I am right we should omit it.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
```diff
@@ -578,12 +576,12 @@ template<typename CharacterType> bool ContentSecurityPolicySourceList::parseNonc
         return false;
     if (extensionModeAllowsKeywordsForDirective(m_contentSecurityPolicyModeForExtension, m_directiveName))
-        m_nonces.add(String({ beginNonceValue, buffer.position() }));
+        m_nonces.add(String(beginNonceValue.first(buffer.position() - beginNonceValue.data())));
```
I don’t think we need `String()` here; if I am right we should omit it.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/
https://bugs.webkit.org/show_bug.cgi?id=283201

Reviewed by Darin Adler.

* Source/WebCore/loader/FormSubmission.cpp:
(WebCore::appendMailtoPostFormDataToURL):
* Source/WebCore/loader/ResourceCryptographicDigest.cpp:
(WebCore::parseCryptographicDigestImpl):
(WebCore::parseEncodedCryptographicDigestImpl):
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::shouldDeferUpdateImageData const):
* Source/WebCore/page/PerformanceObserver.cpp:
(WebCore::PerformanceObserver::observe):
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::parseDirective):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::parse):
(WebCore::ContentSecurityPolicySourceList::parseSource):
(WebCore::ContentSecurityPolicySourceList::parseScheme):
(WebCore::ContentSecurityPolicySourceList::parseHost):
(WebCore::ContentSecurityPolicySourceList::parsePath):
(WebCore::ContentSecurityPolicySourceList::parsePort):
(WebCore::ContentSecurityPolicySourceList::parseNonceSource):

Canonical link: https://commits.webkit.org/286696@main
